### DISCUSSION TITLE
I can not get upgrate by embargo

### DISCUSSION THREAD

**Discussion by jyu99** [fwupd/fwupd#2686] [General]
```
03:40:37:0113 FuCommon             reading /home/jimmy_yu/Jimmy/Test/var/lib/fwupd/remotes.d/pixart-embargo/metadata.xml.gz with 1277 bytes
03:40:37:0113 FuCommon             reading /home/jimmy_yu/Jimmy/Test/var/lib/fwupd/remotes.d/pixart-embargo/metadata.xml.gz.jcat with 1240 bytes
03:40:37:0113 Jcat                 got 1 PKCS7 signatures
03:40:37:0113 Jcat                 signature failure: failed to verify data for O=Linux Vendor Firmware Project,CN=LVFS CA: Public key signature verification has failed. [-89]
03:40:37:0116 Jcat                 returned signature fingerprint 3FC6B804410ED0840D8F2F9748A6D80E4538BAC2
03:40:37:0116 Jcat                 verified: JcatResult:
Timestamp:             2020-12-14T07:06:54Z
Authority:             3FC6B804410ED0840D8F2F9748A6D80E4538BAC2
JcatGpgEngine:
Kind:                gpg
VerifyKind:          signature

03:40:37:0117 FuCommon             writing /home/jimmy_yu/Jimmy/Test/var/lib/fwupd/remotes.d/pixart-embargo/metadata.xml.gz with 1277 bytes
03:40:37:0117 FuCommon             writing /home/jimmy_yu/Jimmy/Test/var/lib/fwupd/remotes.d/pixart-embargo/metadata.xml.gz.jcat with 1240 bytes
03:40:37:0117 XbSilo               attempting to load /home/jimmy_yu/Jimmy/Test/var/cache/fwupd/metadata.xmlb
03:40:37:0117 XbSilo               file: 84756e79-0fdb-cc5c-32f1-f5feba944608, current:9520ac98-4cf5-577a-0393-2cef2fce351e, cached: (null)
03:40:37:0117 XbSilo               compiling /home/jimmy_yu/Jimmy/Test/var/lib/fwupd/remotes.d/pixart-embargo/metadata.xml.gz:ctime=1608003637.116785:func-id=AppStreamUpgrade@3:value=/home/jimmy_yu/Jimmy/Test/var/lib/fwupd/remotes.d/pixart-embargo/metadata.xml.gz:value=pixart-embargo…
03:40:37:0117 XbSilo               compiling /home/jimmy_yu/Jimmy/Test/share/fwupd/remotes.d/dell-esrt/metadata.xml:ctime=1608001533.300244:func-id=AppStreamUpgrade@3:value=/home/jimmy_yu/Jimmy/Test/share/fwupd/remotes.d/dell-esrt/metadata.xml:value=dell-esrt…
03:40:37:0117 XbSilo               compiling /home/jimmy_yu/Jimmy/Test/var/lib/fwupd/remotes.d/lvfs/metadata.xml.gz:ctime=1608003598.815897:func-id=AppStreamUpgrade@3:value=/home/jimmy_yu/Jimmy/Test/var/lib/fwupd/remotes.d/lvfs/metadata.xml.gz:value=lvfs…
03:40:37:0220 FuEngine             505 components now in silo
```
```

├─Embargoed for pixart:
│     Remote ID:          pixart-embargo
│     Type:               download
│     Keyring:            jcat
│     Enabled:            true
│     Checksum:           df96e07f787e16f8074052a5662d3571a72ac56f50fa0977da78a5ba1f95a183
│     Age:                3.08m
│     Priority:           1
│     Filename:           /home/jimmy_yu/Jimmy/Test/var/lib/fwupd/remotes.d/pixart-embargo/metadata.xml.gz
│     Filename Signature: /home/jimmy_yu/Jimmy/Test/var/lib/fwupd/remotes.d/pixart-embargo/metadata.xml.gz.jcat
│     Metadata URI:       https://fwupd.org/downloads/firmware-fddc222bd7a6146f4c4f63296ec6c956365cf171837ec34fc92a5c9c1d7d2082.xml.gz
│     Metadata Signature: https://fwupd.org/downloads/firmware-fddc222bd7a6146f4c4f63296ec6c956365cf171837ec34fc92a5c9c1d7d2082.xml.gz.jcat
│     Report URI:         https://fwupd.org/lvfs/firmware/report
│     Automatic Reporting:false
```

I had done local vendor-directory. It's all fine for upgrated. Then I try to use pixart-embargo. I can't get  the upgrade info
I guess it cause by the pgp verify fail.

Does anyone get this problem before?

**Comment by superm1** [comment]
What host OS?  Is this hand compiled fwupd and jcat?

**Reply by jyu99** [reply]
Hi here is the OS info
```
> NAME=Fedora
> VERSION="33 (Workstation Edition)"
> ID=fedora
> VERSION_ID=33
> VERSION_CODENAME=""
> PLATFORM_ID="platform:f33"
```
I use the hand complied fwupd. All operation is follow fwupd wiki fedora install steps 
https://github.com/fwupd/fwupd/wiki/Fedora-31

**Reply by superm1** [reply]
@hughsie to me this seems like it's likely an issue with the embargo channels.  I've grabbed the Dell one and tried to verify using `jcat-tool` and see something similar (files renamed)

```
jcat-tool verify foo.xml.gz.jcat 
foo.xml.gz:
    FAILED sha1: verifying data is not supported
    FAILED sha256: verifying data is not supported
    FAILED pkcs7: failed to verify data for O=Linux Vendor Firmware Project,CN=LVFS CA: Public key signature verification has failed. [-89]
    PASSED gpg: 3FC6B804410ED0840D8F2F9748A6D80E4538BAC2
    FAILED: Validation failed
Validation failed
```

**Reply by hughsie** [reply]
> this seems like it's likely an issue with the embargo channels

Not convinced; I can see the same thing with `jcat-tool verify firmware.xml.gz.jcat --public-keys /etc/pki/fwupd-metadata/ --verbose` -- there's something funky with the PKCS7 signing or verification. This shouldn't affect fwupd as it just needs either signature type to pass to accept the metadata.

**Reply by hughsie** [reply]
So, there's something funky with the signing on the LVFS; switch it back to `certtool --p7-detached-sign` and it verifies fine (even in PyGnuTLS). Debugging continues....

**Reply by hughsie** [reply]
So the problem was that the PyGnuTLS code was not saving the certificate to the signature -- I've fixed that in https://gitlab.com/fwupd/lvfs-website/-/commit/6659b6b23bfb4589e5359490000d34bbde5caa70 -- just one added line but it took most of today to track down!

With the regenerated admin metadata I now get:

    PASSED pkcs7: O=Linux Vendor Firmware Project,CN=LVFS CA
    PASSED gpg: 3FC6B804410ED0840D8F2F9748A6D80E4538BAC2

for `jcat-tool verify firmware.xml.gz.jcat --public-keys /etc/pki/fwupd-metadata/` -- I'll resign all the metadata remotes now, and that will take a few minutes to complete.

**Reply by hughsie** [reply]
@jyu99 I've resigned `pixart` just now, effective immediate -- same for @superm1 with `dell`. All the others will regenerate in ~3 hours.

**Reply by superm1** [reply]
Pixart probably needs a new URL since this one was leaked...

**Reply by hughsie** [reply]
I think this might be the right time to think about either username/password or username/token for the download of the per-vendor metadata... Agreed?

**Reply by superm1** [reply]
I think between those two I would rather see revokable tokens tied to a user login.

**Comment by hughsie** [comment]
@jyu99 can you also fix the problem specified in https://fwupd.org/lvfs/firmware/6053/problems -- I think that's actually the issue here (although the PKCS thing needed fixing anyway!) -- and with the update protocol set it should just work.

**Reply by jyu99** [reply]
Verifying flow  could not be a cause for embargo downloading fail. After I traced the code, there is something strange in source code.
I noticed the function `fu_engine_add_releases_for_device_component` in `fu-enging.c`

```
static gboolean
fu_engine_add_releases_for_device_component (FuEngine *self,
					     FuEngineRequest *request,
					     FuDevice *device,
					     XbNode *component,
					     GPtrArray *releases,
					     GError **error)
{
	g_debug("fu_engine_add_releases_for_device_component");
	FwupdFeatureFlags feature_flags;
	FwupdVersionFormat fmt = fu_device_get_version_format (device);
	g_autoptr(GError) error_local = NULL;
	g_autoptr(FuInstallTask) task = fu_install_task_new (device, component);
	g_autoptr(GPtrArray) releases_tmp = NULL;

	if (!fu_engine_check_requirements (self, request, task,
					   FWUPD_INSTALL_FLAG_OFFLINE |
					   FWUPD_INSTALL_FLAG_IGNORE_VID_PID |
					   FWUPD_INSTALL_FLAG_ALLOW_BRANCH_SWITCH |
					   FWUPD_INSTALL_FLAG_ALLOW_REINSTALL |
					   FWUPD_INSTALL_FLAG_ALLOW_OLDER,
					   error))
					   {
```
 I can update successfully, if I remove flag `FWUPD_INSTALL_FLAG_IGNORE_VID_PID` parameter from `fu_engine_check_requirements`
I check the release meta data xml file,then it exist the requirement in xml. I guess it added  automatically by LVFS server
below is part of release metadata xml from LVFS

```
   <requires>
      <firmware compare="eq" version="USB:0x093A">vendor-id</firmware>
    </requires>
```


`FWUPD_INSTALL_FLAG_IGNORE_VID_PID` be parameter added on 12/3  commit.

Should I roll-back the other stable release?

below is the verbose info for  passing parameter`FWUPD_INSTALL_FLAG_IGNORE_VID_PID`  into `fu_engine_check_requirements`

```
11:53:12:0737 FuEngine             @@@fu_engine_get_upgrades
11:53:12:0737 FuEngine             fu_engine_get_releases_for_device
11:53:12:0738 XbSilo               1.12ms      query on / with `components/component/provides/firmware[@type=$'flashed'][text()=$'e84daf50-dfb6-5800-807b-1fb4aa54a4a8']/../..|components/component/provides/firmware[@type=$'flashed'][text()=$'7bd6065d-14d7-54a5-9487-428d66c18936']/../..` limit=0 -> 1 results
11:53:12:0738 FuEngine             fu_engine_add_releases_for_device_component
11:53:12:0738 XbSilo               0.03ms      query on component with `provides/firmware[@type='flashed']` limit=0 -> 2 results
11:53:12:0738 XbSilo               0.03ms      query on component with `custom/value[@key='LVFS::UpdateProtocol']` limit=1 -> 1 results
11:53:12:0738 XbSilo               0.01ms      query on component with `releases/release` limit=1 -> 1 results
11:53:12:0738 XbSilo               0.01ms      query on component with `requires/*` limit=0 -> 1 results
11:53:12:0738 FuEngine             fu_engine_check_requirement
11:53:12:0738 FuEngine             name 256 MOUSE flag 167 fu_engine_check_requirement_firmware
11:53:12:0738 FuEngine             xnode vendor-id
11:53:12:0738 FuEngine             bad false
No releases found: cannot handle firmware requirement 'vendor-id'

```

**Reply by hughsie** [reply]
Doesn't d483fcedd5e54b5fc805a9826e8568982c28cc31 already fix this?

**Reply by jyu99** [reply]
Sorry.. I  got the  wrong commit. You're right this bug had fixed

**Reply by jyu99** [reply]
@hughsie thank your help .  I still have a problem. where can I apply the update protocol in LVFS?
or should I upload our plugin to repository in first step?

**Reply by hughsie** [reply]
@jyu99 you need to change the protocol to presumably DFU in https://fwupd.org/lvfs/components/7253

**Reply by jyu99** [reply]
but if I want set unique update protocol, just like "com.pixart.sdrf.hid", where can I apply ?or  I need upload the our plugin via pull request? Let you know our update protocol and then you will help us add into LVFS  ?

**Reply by hughsie** [reply]
> or I need upload the our plugin via pull request

You need to submit your plugin code to fwupd before we can add custom protocols to the LVFS I'm afraid.

**Reply by jyu99** [reply]
Thank your reply @hughsie

**Comment by jyu99** [comment]
Is there Any way to get the release version in `fu_device_write_firmware `function or  `fu_device_prepare_firmware `function ?

**Comment by hughsie** [comment]
> Any way to get the release version

If the firmware can know the version, you can use fu_firmware_get_version() -- but if you need to know the version in the metainfo.xml file please file a new issue and we can try and support that for you.
